### PR TITLE
Fix for cdist backward for non-batch tensors

### PR DESCRIPTION
**Pull request by ifedan**
Fix for: https://github.com/pytorch/pytorch/issues/22353

### CODE REVIEW

**Comment by zou3519** [comment]
What is `cdist`? There doesn't seem to be a documentation page at `torch.cdist` and `doc(torch.cdist)` is None.

**Comment by ifedan** [comment]
> What is `cdist`? There doesn't seem to be a documentation page at `torch.cdist` and `doc(torch.cdist)` is None.

Yep :-(, I will work on this as separate PR. cdist corresponds to https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 1]
```diff
@@ -325,7 +324,10 @@ void cdist_backward_kernel_impl(Tensor& result, const Tensor& grad, const Tensor
   const int64_t r1 = x1.size(-2);
   const int64_t r2 = x2.size(-2);
   const int64_t m = x1.size(-1);
-  const int64_t d = x1.size(0);
+  int64_t d = 1;
```
Comment (zou3519) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: Shorter as a ternary operator, but sometimes less readable: `int64_t d = x1.dim() > 2 ? 1 : x1.size(0)`.

What is `d`? It doesn't have a very descriptive name

File: test/test_autograd.py [hunk 2]
```diff
@@ -2228,16 +2228,22 @@ def test_trapz(self):
...
-                               torch.randn(S, S, requires_grad=True))
-
-            def f(a, b):
-                return torch.cdist(a, b, p)
-
-            f_args_tensor = deepcopy(unpack_variables(f_args_variable))
-            run_functional_checks(self, "test_cdist", "cdist", f,
-                                  True, f_args_variable, f_args_tensor)
+        def _test_cdist_for_size(sizes):
+            devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
```
Comment (zou3519) [inline test/test_autograd.py]: I know we use this idiom a lot but the better way to do this is
`devices = torch.testing.get_all_device_types()`

File: test/test_autograd.py [hunk 3]
```diff
@@ -2228,16 +2228,22 @@ def test_trapz(self):
...
+
+                    def f(a, b):
+                        return torch.cdist(a, b, p)
+
+                    f_args_tensor = deepcopy(unpack_variables(f_args_variable))
+                    run_functional_checks(self, "test_cdist", "cdist", f,
+                                          True, f_args_variable, f_args_tensor)
+
+        _test_cdist_for_size((S, S))
+        _test_cdist_for_size((S, S, S))
```
Comment (gchanan) [inline test/test_autograd.py]: can you mix up the sizes?  There's a bunch of code above that grabs various sizes, but all of the sizes in your test are the same so you wouldn't notice a mistake there.

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
did you file an issue for the missing documentation?

**Comment by ifedan** [comment]
> did you file an issue for the missing documentation?

It's already there:
https://github.com/pytorch/pytorch/issues/23675

**Comment by bddppq** [comment]
autograd results mismatch with ROCm cc @iotamudelta 
```
21:31:18 ======================================================================
21:31:18 ERROR: test_cdist (__main__.TestAutograd)
21:31:18 ----------------------------------------------------------------------
21:31:18 Traceback (most recent call last):
21:31:18   File "test_autograd.py", line 2477, in test_cdist
21:31:18     _test_cdist_for_size((S, S))
21:31:18   File "test_autograd.py", line 2475, in _test_cdist_for_size
21:31:18     True, f_args_variable, f_args_tensor)
21:31:18   File "test_autograd.py", line 3652, in run_functional_checks
21:31:18     output_variable, f_args_variable)
21:31:18   File "test_autograd.py", line 3633, in run_grad_and_gradgrad_checks
21:31:18     test_case.assertTrue(gradcheck(apply_method, input_variables, eps=1e-6, atol=PRECISION))
21:31:18   File "/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 289, in gradcheck
21:31:18     'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
21:31:18   File "/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 227, in fail_test
21:31:18     raise RuntimeError(msg)
21:31:18 RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
21:31:18 numerical:tensor([[-1.0000, -1.0000, -1.0000, -1.0000, -1.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [-0.9984, -0.9984,  0.9835,  1.0133,  1.0729,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [-0.9984, -0.9537, -0.9537,  0.9984, -1.0133,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [-0.9984,  0.9984,  0.9835,  1.0133,  1.0729,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [-0.9835,  0.9984, -0.9537, -0.9835, -0.9984,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  1.0000,  1.0000, -1.0000,
21:31:18           1.0000,  1.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537, -0.9537, -1.0133,
21:31:18           1.0133,  0.9537,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9984, -0.9537, -0.9537,
21:31:18           0.9984, -1.0133,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537, -0.9537, -1.0133,
21:31:18          -1.0133, -0.9537,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537, -0.9984, -1.0133,
21:31:18          -1.0133, -0.9984,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000, -1.0000, -1.0000, -1.0000,  1.0000, -1.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000, -0.9537, -1.0133, -1.0133,  1.0133,  1.0133,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  1.0133, -0.9984, -0.9984,  1.0729,  1.0133,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  1.0133,  1.0133,  0.9835,  1.0133,  1.0133,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000, -1.0133, -0.9984, -0.9537, -1.0133, -0.9984,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -1.0000,
21:31:18          -1.0000, -1.0000,  1.0000, -1.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537,
21:31:18          -1.0133, -1.0133, -1.0133, -1.0133,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -1.0133,
21:31:18          -0.9537, -0.9537,  1.0002, -1.0133,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537,
21:31:18          -1.0133,  1.0133,  1.0133,  1.0133,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000, -0.9537,
21:31:18          -1.0133, -0.9537, -0.9537, -0.9537,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  1.0000, -1.0000, -1.0000,  1.0000,
21:31:18           1.0000],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000, -0.9537, -1.0133,  1.0133,  0.9537,
21:31:18           1.0729],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000, -1.0133, -1.1921, -0.9537, -1.0133,
21:31:18          -1.0729],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.9537,  0.9537,  1.0133,  0.9537,
21:31:18           0.9537],
21:31:18         [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
21:31:18           0.0000,  0.0000,  0.0000,  0.0000,  0.9984,  0.9537, -0.9984, -1.0002,
21:31:18           1.0133]])
21:31:18 analytical:tensor([[-1., -1., -1., -1., -1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [-1., -1.,  1.,  1.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [-1., -1., -1.,  1., -1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [-1.,  1.,  1.,  1.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [-1.,  1., -1., -1., -1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  1.,  1., -1.,  1.,  1.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0., -1., -1., -1.,  1.,  1.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0., -1., -1., -1.,  1., -1.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0., -1., -1., -1., -1., -1.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0., -1., -1., -1., -1., -1.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -1., -1., -1.,  1.,
21:31:18          -1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -1., -1., -1.,  1.,
21:31:18           1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1., -1., -1.,  1.,
21:31:18           1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,
21:31:18           1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -1., -1., -1., -1.,
21:31:18          -1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0., -1., -1., -1.,  1., -1.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0., -1., -1., -1., -1., -1.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0., -1., -1., -1.,  1., -1.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0., -1., -1.,  1.,  1.,  1.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0., -1., -1., -1., -1., -1.,  0.,  0.,  0.,  0.,  0.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  1., -1., -1.,  1.,  1.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0., -1., -1.,  1.,  1.,  1.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0., -1., -1., -1., -1., -1.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,  1.],
21:31:18         [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
21:31:18           0.,  0.,  0.,  0.,  0.,  0.,  1.,  1., -1., -1.,  1.]])
21:31:18 
21:31:18 
21:31:18 ----------------------------------------------------------------------
```

@ifedan To unblock, use the `@skipIfRocm` decorator to skip the test in ROCM CI

**Comment by iotamudelta** [comment]
@ifedan I would assume `__shfl_down` is being used, there is no `__shfl_down_sync` for ROCm AFAIK. I also would assume the error is not because `__shfl_down` gives wrong results but instead that it's being used wrongly, most likely because a warp size of 32 is assumed - all ROCm targets currently have a warp size of 64. Just a note.

**Comment by ifedan** [comment]
> @ifedan I would assume `__shfl_down` is being used, there is no `__shfl_down_sync` for ROCm AFAIK. I also would assume the error is not because `__shfl_down` gives wrong results but instead that it's being used wrongly, most likely because a warp size of 32 is assumed - all ROCm targets currently have a warp size of 64. Just a note.

@iotamudelta I think WARP_SIZE is not used, we use instead warpSize which should have a right size

**Pull request closed without merging**